controllers/helpers.go

```diff
@@ -404,3 +434,73 @@ func GetClusterIdentityFromRef(ctx context.Context, c client.Client, azureCluste
 	}
 	return nil, nil
 }
+
+// CoalescingRequestCache uses and underlying time to live last recently used cache to track high frequency requests.
+// A reconciler should call ShouldProcess to determine if the key has expired. If the key has expired, a zero value
+// time.Time and true is returned. If the key has not expired, the expiration and false is returned. Upon successful
+// reconciliation a reconciler should call Reconciled to update the cache expiry.
+type CoalescingRequestCache struct {
+	lastSuccessfulReconciliationCache *ttllru.Cache
+}
+
+// NewCoalescingRequestCache creates a new instance of a CoalescingRequestCache given a specified window of expiration
+func NewCoalescingRequestCache(window time.Duration) (*CoalescingRequestCache, error) {
+	cache, err := ttllru.New(1024, window)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to build ttllru cache")
+	}
+
+	return &CoalescingRequestCache{
+		lastSuccessfulReconciliationCache: cache,
+	}, nil
+}
+
+// ShouldProcess determines if the key has expired. If the key has expired, a zero value
+// time.Time and true is returned. If the key has not expired, the expiration and false is returned.
+func (cache *CoalescingRequestCache) ShouldProcess(key string) (time.Time, bool) {
+	_, expiration, ok := cache.lastSuccessfulReconciliationCache.Peek(key)
+	return expiration, !ok
+}
+
+// Reconciled updates the cache expiry for a given key
+func (cache *CoalescingRequestCache) Reconciled(key string) {
+	cache.lastSuccessfulReconciliationCache.Add(key, nil)
+}
+
+type coalescingReconciler struct {
+	upstream reconcile.Reconciler
+	cache    *CoalescingRequestCache
+	log      logr.Logger
+}
+
+// NewCoalescingReconciler returns a reconcile wrapper that will delay new reconcile.Requests
+// after the cache expiry of the request string key.
+// A successful reconciliation is defined as as one where no error is returned
+func NewCoalescingReconciler(upstream reconcile.Reconciler, cache *CoalescingRequestCache, log logr.Logger) reconcile.Reconciler {
```

Comment:
This wraps the `AzureMachinePool` and `AzureMachinePoolMachine` reconcilers so that they debounce, the reconcilers rate limit the incoming events so they only do so many within a window of time to not overwhelm Azure API limits.

There is no good way to do this in controller-runtime. I'll intro a proposal over there for middleware or incoming rate limiting.